### PR TITLE
Checks if deck name already exists before creating a new deck.

### PR DESCRIPTION
**Pull request by gourabsingha1**
## Purpose / Description
The purpose of this change is to check if the deck name already exists before creating a new deck. This ensures that no deck with the same name exists.

## Fixes
* Fixes https://github.com/ankidroid/Anki-Android/issues/15678

## Approach
Created an object which stores the deck list when it is rendered. Added a check function which checks if the deck name already exists in the list before creating a new deck.

## How Has This Been Tested?
The changes have been tested thoroughly to ensure functionality and compatibility across different devices and screen sizes. The following steps were taken to verify the changes:

1. Opened the "Create deck" section
2. Entered "Default" as the first deck. It didn't create since "Default" deck already exists.
3. Created "test" deck.
4. Tried to create "test" deck again. But toast popped up notifying that the deck name already exists.
5. Reviewed the code changes for any potential issues or regressions.

## Learning (optional, can help others)
During the implementation of this feature, I learned about creating an object that is accessible from all files.

No additional external resources were used in the implementation of this feature.

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by david-allison — CHANGES_REQUESTED** [review]
Hi, CI is failing

Please only add strings to `/values`, our translation process takes care of the rest

https://github.com/ankidroid/Anki-Android/wiki/Development-Guide#adding-translations

**Comment by gourabsingha1** [comment]
Help me please. This is my first pull request.

File: build.gradle [hunk 1]
```diff
@@ -109,21 +109,21 @@ subprojects {
...
-        System.exit(1)
-    }
+   if (jvmVersion != "17" && jvmVersion != "21") {
+       println "\n\n\n"
+       println "**************************************************************************************************************"
+       println "\n\n\n"
+       println "ERROR: AnkiDroid builds with JVM version 17 or 21."
+       println "  Incompatible major version detected: '" + jvmVersion + "'"
+       println "\n\n\n"
+       println "  If you receive this error because you want to use a newer JDK, we may accept PRs to support new versions."
```
Comment (criticalAY) [inline build.gradle:119]: lint takes care of it

File: AnkiDroid/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string name="deck_name_already_exists">Deck name already exists</string>
+</resources>
```
Comment (criticalAY) [inline AnkiDroid/src/main/res/values/strings.xml:4]: As told by David, string changes only in `values/`

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 3]
```diff
@@ -240,9 +242,6 @@ open class DeckPicker :
     @VisibleForTesting
     var optionsMenuState: OptionsMenuState? = null
 
-    @VisibleForTesting
-    var dueTree: DeckNode? = null
-
```
Comment (criticalAY) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:245]: Why this?

**Review by criticalAY — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt (hunk 3 above)
Comment (gourabsingha1) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:245]: I've created a DueTree object where I'm keeping this variable for using in any file. Lemme try again without removing it.

**Review by david-allison — CHANGES_REQUESTED** [review]
See: https://github.com/ankidroid/Anki-Android/commit/7a65160e0e23e20080091a30abdd4c05fc610e06#diff-50f072649824bf3954a004d011607850a00599ca5814aaa20b6adbacd044c990

`CreateDeckDialog.kt`

for how this was done previously, both the UX and functionality

**Review by criticalAY — CHANGES_REQUESTED** [review]

**Comment by gourabsingha1** [comment]
Reverted all commits. Please provide me with the patch. I really wanna resolve this.

**Comment by criticalAY** [comment]
You added more commits, could rebase the PR?

**Comment by gourabsingha1** [comment]
I'm so confused

**Comment by gourabsingha1** [comment]
@criticalAY can you please tell me what I'm doing wrong?
This is my first ever contribution. I don't wanna lose motivation.

**Comment by criticalAY** [comment]
You, need to rebase the PR, I can tell you how to do it but since you have closed the PR now, but here is the patch you will have to update the existing test otherwise the CI will fail again
Subject: [PATCH] deck name issue
---
Index: AnkiDroid/src/main/res/layout/dialog_generic_text_input.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/res/layout/dialog_generic_text_input.xml b/AnkiDroid/src/main/res/layout/dialog_generic_text_input.xml
--- a/AnkiDroid/src/main/res/layout/dialog_generic_text_input.xml	(revision c25fd25196699388914a48f64d36d8fa33473d49)
+++ b/AnkiDroid/src/main/res/layout/dialog_generic_text_input.xml	(date 1709823798468)
@@ -15,8 +15,8 @@
   -->
 
 
-<com.google.android.material.textfield.TextInputLayout
-    xmlns:android="http://schemas.android.com/apk/res/android"
+<com.google.android.material.textfield.TextInputLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
     android:id="@+id/dialog_text_input_layout"
     android:layout_width="match_parent"
     android:layout_height="wrap_content"
@@ -24,7 +24,8 @@
     android:paddingLeft="32dp"
     android:paddingRight="32dp"
     android:paddingBottom="4dp"
-    android:paddingTop="16dp" >
+    app:errorEnabled="true"
+    android:paddingTop="16dp">
 
     <com.google.android.material.textfield.TextInputEditText
         android:id="@+id/dialog_text_input"
Index: AnkiDroid/src/main/res/values/03-dialogs.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/res/values/03-dialogs.xml b/AnkiDroid/src/main/res/values/03-dialogs.xml
--- a/AnkiDroid/src/main/res/values/03-dialogs.xml	(revision c25fd25196699388914a48f64d36d8fa33473d49)
+++ b/AnkiDroid/src/main/res/values/03-dialogs.xml	(date 1709824312043)
@@ -260,4 +260,6 @@
         Due to Android privacy changes, your data and automated backups will be inaccessible if the app is uninstalled
     </string>
 
+    <string name="deck_name_exist">Deck name already exist</string>
+
 </resources>
Index: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CreateDeckDialog.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CreateDeckDialog.kt b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CreateDeckDialog.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CreateDeckDialog.kt	(revision c25fd25196699388914a48f64d36d8fa33473d49)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CreateDeckDialog.kt	(date 1709824312049)
@@ -26,10 +26,12 @@
 import com.ichi2.anki.UIUtils.showThemedToast
 import com.ichi2.anki.snackbar.showSnackbar
 import com.ichi2.annotations.NeedsTest
+import com.ichi2.libanki.Collection
 import com.ichi2.libanki.DeckId
 import com.ichi2.libanki.Decks
 import com.ichi2.libanki.getOrCreateFilteredDeck
 import com.ichi2.utils.getInputField
+import com.ichi2.utils.getInputTextLayout
 import com.ichi2.utils.input
 import com.ichi2.utils.negativeButton
 import com.ichi2.utils.positiveButton
@@ -87,6 +89,7 @@
             negativeButton(R.string.dialog_cancel)
             setView(R.layout.dialog_generic_text_input)
         }.input(prefill = initialDeckName, displayKeyboard = true, waitForPositiveButton = false) { dialog, text ->
+
             // we need the fully-qualified name for subdecks
             val maybeDeckName = fullyQualifyDeckName(dialogText = text)
             // if the name is empty, it seems distracting to show an error
@@ -94,12 +97,23 @@
                 dialog.positiveButton.isEnabled = false
                 return@input
             }
+            if (deckExists(getColUnsafe, maybeDeckName)){
+                dialog.getInputTextLayout().error = context.getString(R.string.deck_name_exist)
+                dialog.positiveButton.isEnabled = false
+                return@input
+            } else dialog.getInputTextLayout().error = null
             dialog.positiveButton.isEnabled = true
         }
         shownDialog = dialog
         return dialog
     }
 
+    /**
+     * @return true if the collection contains a deck with the given name
+     */
+    fun deckExists(col: Collection, name: String) =
+        col.decks.byName(name) != null
+
     /**
      * Returns the fully qualified deck name for the provided input
      * @param dialogText The user supplied text in the dialog

**Comment by gourabsingha1** [comment]
Thanks a lot. Now I know that I had to show the error before allowing to create a deck. Not show it after the creation.
I've created another pull request https://github.com/ankidroid/Anki-Android/pull/15808
But lint debug is failing there for some reason. Can you check?

**Comment by criticalAY** [comment]
I already told you that you will have to update the test

**Comment by gourabsingha1** [comment]
My bad. You're right. Its taking time for me to get familiar

**Pull request closed without merging**